File: src/lib/db/repositories/synergyRepository.ts
import { and, eq, inArray, sql } from "drizzle-orm";

import type { AppDatabase } from "@/lib/db/client";
import { synergies, synergyLinks } from "@/lib/db/schema";
import type { SynergyLinkInput } from "@/lib/synergies/schemas";
import type { SynergyType } from "@/lib/synergies/schemas";

export type SynergyRecord = {
  id: string;
  userId: number;
  name: string;
  type: SynergyType;
  subType: string | null;
  description: string;
  createdAt: string;
  updatedAt: string;
};

export type SynergyLinkRecord = {
  id: string;
  synergyId: string;
  kind: string;
  displayName: string;
  itemHash: number | null;
  perkHash: number | null;
  parentItemHash: number | null;
  originTraitName: string | null;
  originTraitHash: number | null;
  armorSetName: string | null;
  bonusPieces: number | null;
  bonusName: string | null;
  armorSetHash: number | null;
  /** Required evidence link (DBR-SYN-007–010a). Omitted/false = soft evidence. */
  required?: boolean;
};

export type SynergyWithLinks = SynergyRecord & { links: SynergyLinkRecord[] };

export type SynergyTargetQuery = {
  kind: string;
  name?: string;
  itemHash?: number;
  perkHash?: number;
  originTraitHash?: number;
  armorSetName?: string;
  bonusPieces?: number;
  bonusName?: string;
};

function rowToSynergy(row: typeof synergies.$inferSelect): SynergyRecord {
  return {
    id: row.id,
    userId: row.userId,
    name: row.name,
    type: row.type as SynergyType,
    subType: row.subType ?? null,
    description: row.description,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  };
}

function mapLinkRow(row: typeof synergyLinks.$inferSelect): SynergyLinkRecord {
  return {
    id: row.id,
    synergyId: row.synergyId,
    kind: row.kind,
    displayName: row.displayName,
    itemHash: row.itemHash ?? null,
    perkHash: row.perkHash ?? null,
    parentItemHash: row.parentItemHash ?? null,
    originTraitName: row.originTraitName ?? null,
    originTraitHash: row.originTraitHash ?? null,
    armorSetName: row.armorSetName ?? null,
    bonusPieces: row.bonusPieces ?? null,
    bonusName: row.bonusName ?? null,
    armorSetHash: row.armorSetHash ?? null,
    required: Boolean(row.required),
  };
}

function listSynergyLinks(db: AppDatabase, synergyId: string): SynergyLinkRecord[] {
  return db
    .select()
    .from(synergyLinks)
    .where(eq(synergyLinks.synergyId, synergyId))
    .all()
    .map(mapLinkRow);
}

/** Batch-load links for many synergies (one query). */
function listSynergyLinksForIds(
  db: AppDatabase,
  synergyIds: string[],
): Map<string, SynergyLinkRecord[]> {
  const map = new Map<string, SynergyLinkRecord[]>();
  for (const id of synergyIds) map.set(id, []);
  if (synergyIds.length === 0) return map;

  const rows = db
    .select()
    .from(synergyLinks)
    .where(inArray(synergyLinks.synergyId, synergyIds))
    .all();

  for (const row of rows) {
    const mapped = mapLinkRow(row);
    const list = map.get(row.synergyId);
    if (list) list.push(mapped);
    else map.set(row.synergyId, [mapped]);
  }
  return map;
}

function rowsToSynergiesWithLinks(
  db: AppDatabase,
  rows: (typeof synergies.$inferSelect)[],
): SynergyWithLinks[] {
  if (rows.length === 0) return [];
  const linksBySynergy = listSynergyLinksForIds(
    db,
    rows.map((r) => r.id),
  );
  return rows.map((row) => ({
    ...rowToSynergy(row),
    links: linksBySynergy.get(row.id) ?? [],
  }));
}

function insertLinks(db: AppDatabase, synergyId: string, links: SynergyLinkInput[]): void {
  for (const link of links) {
    db.insert(synergyLinks)
      .values({
        id: crypto.randomUUID(),
        synergyId,
        kind: link.kind,
        displayName: link.displayName,
        itemHash: link.itemHash ?? null,
        perkHash: link.perkHash ?? null,
        parentItemHash: link.parentItemHash ?? null,
        originTraitName: link.originTraitName ?? null,
        originTraitHash: link.originTraitHash ?? null,
        armorSetName: link.armorSetName ?? null,
        bonusPieces: link.bonusPieces ?? null,
        bonusName: link.bonusName ?? null,
        armorSetHash: link.armorSetHash ?? null,
        required: link.required === true ? 1 : 0,
      })
      .run();
  }
}

export function listSynergies(db: AppDatabase, userId: number, type?: SynergyType): SynergyWithLinks[] {
  const rows = db
    .select()
    .from(synergies)
    .where(type ? and(eq(synergies.userId, userId), eq(synergies.type, type)) : eq(synergies.userId, userId))
    .all();
  return rowsToSynergiesWithLinks(db, rows);
}

export function getSynergy(db: AppDatabase, userId: number, id: string): SynergyWithLinks | null {
  const row = db
    .select()
    .from(synergies)
    .where(and(eq(synergies.id, id), eq(synergies.userId, userId)))
    .get();
  if (!row) return null;
  return { ...rowToSynergy(row), links: listSynergyLinks(db, row.id) };
}

export function getSynergiesByIds(db: AppDatabase, userId: number, ids: string[]): SynergyWithLinks[] {
  if (ids.length === 0) return [];
  const rows = db
    .select()
    .from(synergies)
    .where(and(eq(synergies.userId, userId), inArray(synergies.id, ids)))
    .all();
  return rowsToSynergiesWithLinks(db, rows);
}

/** Match library synergies by type + subType (null-safe). */
export function getSynergiesByTypeSubType(
  db: AppDatabase,
  userId: number,
  type: SynergyType,
  subType: string | null,
): SynergyWithLinks[] {
  const rows = db
    .select()
    .from(synergies)
    .where(and(eq(synergies.userId, userId), eq(synergies.type, type)))
    .all();
  const normalizedSub = subType?.trim() || null;
  const matched = rows.filter((row) => {
    const rowSub = row.subType?.trim() || null;
    return rowSub === normalizedSub;
  });
  return rowsToSynergiesWithLinks(db, matched);
}

export function createSynergyRecord(
  db: AppDatabase,
  userId: number,
  input: {
    id: string;
    name: string;
    type: SynergyType;
    subType: string | null;
    description: string;
    links: SynergyLinkInput[];
    now: string;
  },
): SynergyWithLinks {
  db.insert(synergies)
    .values({
      id: input.id,
      userId,
      name: input.name,
      type: input.type,
      subType: input.subType,
      description: input.description,
      createdAt: input.now,
      updatedAt: input.now,
    })
    .run();

  insertLinks(db, input.id, input.links);
  return getSynergy(db, userId, input.id)!;
}

export function updateSynergyRecord(
  db: AppDatabase,
  userId: number,
  id: string,
  patch: {
    name?: string;
    type?: SynergyType;
    subType?: string | null;
    description?: string;
    links?: SynergyLinkInput[];
    now: string;
  },
): SynergyWithLinks | null {
  const existing = getSynergy(db, userId, id);
  if (!existing) return null;

  db.update(synergies)
    .set({
      name: patch.name ?? existing.name,
      type: patch.type ?? existing.type,
      subType: patch.subType !== undefined ? patch.subType : existing.subType,
      description: patch.description ?? existing.description,
      updatedAt: patch.now,
    })
    .where(and(eq(synergies.id, id), eq(synergies.userId, userId)))
    .run();

  if (patch.links) {
    db.delete(synergyLinks).where(eq(synergyLinks.synergyId, id)).run();
    insertLinks(db, id, patch.links);
  }

  return getSynergy(db, userId, id);
}

export function deleteSynergyRecord(db: AppDatabase, userId: number, id: string): boolean {
  const result = db
    .delete(synergies)
    .where(and(eq(synergies.id, id), eq(synergies.userId, userId)))
    .run();
  return result.changes > 0;
}

export function findSynergiesByTarget(
  db: AppDatabase,
  userId: number,
  query: SynergyTargetQuery,
): SynergyWithLinks[] {
  const conditions = [eq(synergies.userId, userId), eq(synergyLinks.kind, query.kind)];

  if (query.itemHash !== undefined) {
    conditions.push(eq(synergyLinks.itemHash, query.itemHash));
  }
  if (query.perkHash !== undefined) {
    conditions.push(eq(synergyLinks.perkHash, query.perkHash));
  }
  if (query.originTraitHash !== undefined) {
    conditions.push(eq(synergyLinks.originTraitHash, query.originTraitHash));
  }
  if (query.name) {
    conditions.push(
      sql`lower(${synergyLinks.originTraitName}) = lower(${query.name})`,
    );
  }
  if (query.armorSetName) {
    conditions.push(
      sql`lower(${synergyLinks.armorSetName}) = lower(${query.armorSetName})`,
    );
  }
  if (query.bonusPieces !== undefined) {
    conditions.push(eq(synergyLinks.bonusPieces, query.bonusPieces));
  }
  if (query.bonusName) {
    conditions.push(sql`lower(${synergyLinks.bonusName}) = lower(${query.bonusName})`);
  }

  const synergyIds = db
    .select({ synergyId: synergyLinks.synergyId })
    .from(synergyLinks)
    .innerJoin(synergies, eq(synergyLinks.synergyId, synergies.id))
    .where(and(...conditions))
    .all()
    .map((r) => r.synergyId);

  const unique = [...new Set(synergyIds)];
  return getSynergiesByIds(db, userId, unique);
}

/**
 * Batch reverse-lookup for item-hash link kinds (weapon, exotic_armor, …).
 * Returns a map keyed by itemHash string → distinct synergies for that hash.
 */
export function findSynergiesByItemHashes(
  db: AppDatabase,
  userId: number,
  kind: string,
  itemHashes: number[],
): Map<number, SynergyWithLinks[]> {
  const uniqueHashes = [...new Set(itemHashes.filter((h) => Number.isFinite(h)))];
  const result = new Map<number, SynergyWithLinks[]>();
  for (const h of uniqueHashes) result.set(h, []);
  if (uniqueHashes.length === 0) return result;

  const rows = db
    .select({
      synergyId: synergyLinks.synergyId,
      itemHash: synergyLinks.itemHash,
    })
    .from(synergyLinks)
    .innerJoin(synergies, eq(synergyLinks.synergyId, synergies.id))
    .where(
      and(
        eq(synergies.userId, userId),
        eq(synergyLinks.kind, kind),
        inArray(synergyLinks.itemHash, uniqueHashes),
      ),
    )
    .all();

  const synergyIds = [...new Set(rows.map((r) => r.synergyId))];
  const synergiesById = new Map(
    getSynergiesByIds(db, userId, synergyIds).map((s) => [s.id, s] as const),
  );

  const seenPerHash = new Map<number, Set<string>>();
  for (const row of rows) {
    if (row.itemHash == null) continue;
    const synergy = synergiesById.get(row.synergyId);
    if (!synergy) continue;
    let seen = seenPerHash.get(row.itemHash);
    if (!seen) {
      seen = new Set();
      seenPerHash.set(row.itemHash, seen);
    }
    if (seen.has(synergy.id)) continue;
    seen.add(synergy.id);
    result.get(row.itemHash)!.push(synergy);
  }

  return result;
}

export function seedDefaultSynergies(db: AppDatabase, userId: number): SynergyWithLinks[] {
  const existing = listSynergies(db, userId);
  if (existing.length > 0) return existing;

  const now = new Date().toISOString();
  createSynergyRecord(db, userId, {
    id: crypto.randomUUID(),
    name: "Melee Combo",
    type: "melee",
    subType: "Base",
    description: "Default melee synergy for dev/testing",
    links: [],
    now,
  });

  return listSynergies(db, userId);
}

File: flutter/packages/db/lib/src/repos/synergy_repository.dart
import 'package:drift/drift.dart';

import '../app_database.dart';
import '../tables.dart';
import 'library_records.dart';

SynergyRecord _rowToSynergy(Synergy row) {
  return SynergyRecord(
    id: row.id,
    userId: row.userId,
    name: row.name,
    type: row.type,
    subType: row.subType,
    description: row.description,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  );
}

SynergyLinkRecord _rowToLink(SynergyLink row) {
  return SynergyLinkRecord(
    id: row.id,
    synergyId: row.synergyId,
    kind: row.kind,
    displayName: row.displayName,
    itemHash: row.itemHash,
    perkHash: row.perkHash,
    parentItemHash: row.parentItemHash,
    originTraitName: row.originTraitName,
    originTraitHash: row.originTraitHash,
    armorSetName: row.armorSetName,
    bonusPieces: row.bonusPieces,
    bonusName: row.bonusName,
    armorSetHash: row.armorSetHash,
    required: row.required != 0,
  );
}

Future<List<SynergyLinkRecord>> _listLinks(
  AppDatabase db,
  String synergyId,
) async {
  final rows = await (db.select(db.synergyLinks)
        ..where((t) => t.synergyId.equals(synergyId)))
      .get();
  return rows.map(_rowToLink).toList();
}

Future<Map<String, List<SynergyLinkRecord>>> _listLinksForIds(
  AppDatabase db,
  List<String> synergyIds,
) async {
  final map = {for (final id in synergyIds) id: <SynergyLinkRecord>[]};
  if (synergyIds.isEmpty) return map;
  final rows = await (db.select(db.synergyLinks)
        ..where((t) => t.synergyId.isIn(synergyIds)))
      .get();
  for (final row in rows) {
    map.putIfAbsent(row.synergyId, () => <SynergyLinkRecord>[]).add(
          _rowToLink(row),
        );
  }
  return map;
}

Future<List<SynergyWithLinks>> _rowsToSynergies(
  AppDatabase db,
  List<Synergy> rows,
) async {
  if (rows.isEmpty) return const [];
  final linksBy = await _listLinksForIds(db, rows.map((r) => r.id).toList());
  return rows
      .map(
        (row) => SynergyWithLinks(
          record: _rowToSynergy(row),
          links: linksBy[row.id] ?? const [],
        ),
      )
      .toList();
}

/// Input for inserting a synergy link (id optional — generated if omitted).
class SynergyLinkInput {
  const SynergyLinkInput({
    this.id,
    required this.kind,
    required this.displayName,
    this.itemHash,
    this.perkHash,
    this.parentItemHash,
    this.originTraitName,
    this.originTraitHash,
    this.armorSetName,
    this.bonusPieces,
    this.bonusName,
    this.armorSetHash,
    this.required = false,
  });

  final String? id;
  final String kind;
  final String displayName;
  final int? itemHash;
  final int? perkHash;
  final int? parentItemHash;
  final String? originTraitName;
  final int? originTraitHash;
  final String? armorSetName;
  final int? bonusPieces;
  final String? bonusName;
  final int? armorSetHash;
  final bool required;
}

Future<void> _insertLinks(
  AppDatabase db,
  String synergyId,
  List<SynergyLinkInput> links,
) async {
  var i = 0;
  for (final link in links) {
    final linkId = link.id ?? '${synergyId}-link-${i++}';
    await db.into(db.synergyLinks).insert(
          SynergyLinksCompanion.insert(
            id: linkId,
            synergyId: synergyId,
            kind: link.kind,
            displayName: link.displayName,
            itemHash: Value(link.itemHash),
            perkHash: Value(link.perkHash),
            parentItemHash: Value(link.parentItemHash),
            originTraitName: Value(link.originTraitName),
            originTraitHash: Value(link.originTraitHash),
            armorSetName: Value(link.armorSetName),
            bonusPieces: Value(link.bonusPieces),
            bonusName: Value(link.bonusName),
            armorSetHash: Value(link.armorSetHash),
            required: Value(link.required ? 1 : 0),
          ),
        );
  }
}

Future<List<SynergyWithLinks>> listSynergies(
  AppDatabase db,
  int userId, {
  String? type,
}) async {
  final query = db.select(db.synergies)..where((t) => t.userId.equals(userId));
  if (type != null) {
    query.where((t) => t.type.equals(type));
  }
  final rows = await query.get();
  return _rowsToSynergies(db, rows);
}

Future<SynergyWithLinks?> getSynergy(
  AppDatabase db,
  int userId,
  String id,
) async {
  final row = await (db.select(db.synergies)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .getSingleOrNull();
  if (row == null) return null;
  return SynergyWithLinks(
    record: _rowToSynergy(row),
    links: await _listLinks(db, row.id),
  );
}

Future<SynergyWithLinks> createSynergyRecord(
  AppDatabase db,
  int userId, {
  required String id,
  required String name,
  required String type,
  String? subType,
  String description = '',
  List<SynergyLinkInput> links = const [],
  required String now,
}) async {
  await db.into(db.synergies).insert(
        SynergiesCompanion.insert(
          id: id,
          userId: userId,
          name: name,
          type: type,
          subType: Value(subType),
          description: Value(description),
          createdAt: now,
          updatedAt: now,
        ),
      );
  await _insertLinks(db, id, links);
  return (await getSynergy(db, userId, id))!;
}

Future<SynergyWithLinks?> updateSynergyRecord(
  AppDatabase db,
  int userId,
  String id, {
  String? name,
  String? type,
  Value<String?> subType = const Value.absent(),
  String? description,
  List<SynergyLinkInput>? links,
  required String now,
}) async {
  final existing = await getSynergy(db, userId, id);
  if (existing == null) return null;

  await (db.update(db.synergies)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .write(
    SynergiesCompanion(
      name: Value(name ?? existing.name),
      type: Value(type ?? existing.type),
      subType: subType.present ? subType : Value(existing.subType),
      description: Value(description ?? existing.description),
      updatedAt: Value(now),
    ),
  );

  if (links != null) {
    await (db.delete(db.synergyLinks)..where((t) => t.synergyId.equals(id)))
        .go();
    await _insertLinks(db, id, links);
  }
  return getSynergy(db, userId, id);
}

Future<bool> deleteSynergyRecord(
  AppDatabase db,
  int userId,
  String id,
) async {
  final n = await (db.delete(db.synergies)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .go();
  return n > 0;
}

/// Target query for reverse synergy lookup (BR-SYN-004 / GAP-UI-SYN-03).
class SynergyTargetQuery {
  const SynergyTargetQuery({
    required this.kind,
    this.itemHash,
    this.perkHash,
    this.originTraitHash,
    this.name,
    this.armorSetName,
    this.bonusPieces,
    this.bonusName,
  });

  final String kind;
  final int? itemHash;
  final int? perkHash;
  final int? originTraitHash;

  /// Origin trait name (case-insensitive match on originTraitName).
  final String? name;
  final String? armorSetName;
  final int? bonusPieces;
  final String? bonusName;
}

/// Reverse-lookup synergies linked to a single evidence target.
Future<List<SynergyWithLinks>> findSynergiesByTarget(
  AppDatabase db,
  int userId,
  SynergyTargetQuery query,
) async {
  final linkRows = await (db.select(db.synergyLinks)
        ..where((t) => t.kind.equals(query.kind)))
      .get();

  final synergyIds = <String>{};
  for (final row in linkRows) {
    if (query.itemHash != null && row.itemHash != query.itemHash) continue;
    if (query.perkHash != null && row.perkHash != query.perkHash) continue;
    if (query.originTraitHash != null &&
        row.originTraitHash != query.originTraitHash) {
      continue;
    }
    if (query.name != null && query.name!.trim().isNotEmpty) {
      final n = row.originTraitName?.toLowerCase() ?? '';
      if (n != query.name!.trim().toLowerCase()) continue;
    }
    if (query.armorSetName != null && query.armorSetName!.trim().isNotEmpty) {
      final n = row.armorSetName?.toLowerCase() ?? '';
      if (n != query.armorSetName!.trim().toLowerCase()) continue;
    }
    if (query.bonusPieces != null && row.bonusPieces != query.bonusPieces) {
      continue;
    }
    if (query.bonusName != null && query.bonusName!.trim().isNotEmpty) {
      final n = row.bonusName?.toLowerCase() ?? '';
      if (n != query.bonusName!.trim().toLowerCase()) continue;
    }
    synergyIds.add(row.synergyId);
  }

  if (synergyIds.isEmpty) return const [];

  final rows = await (db.select(db.synergies)
        ..where((t) => t.userId.equals(userId) & t.id.isIn(synergyIds.toList())))
      .get();
  return _rowsToSynergies(db, rows);
}

/// Batch reverse-lookup for item-hash link kinds (weapon, exotic_armor, …).
///
/// Returns map keyed by itemHash → distinct synergies for that hash.
Future<Map<int, List<SynergyWithLinks>>> findSynergiesByItemHashes(
  AppDatabase db,
  int userId,
  String kind,
  List<int> itemHashes,
) async {
  final unique = [...{...itemHashes.where((h) => h != 0)}];
  final result = <int, List<SynergyWithLinks>>{
    for (final h in unique) h: <SynergyWithLinks>[],
  };
  if (unique.isEmpty) return result;

  final linkRows = await (db.select(db.synergyLinks)
        ..where(
          (t) => t.kind.equals(kind) & t.itemHash.isIn(unique),
        ))
      .get();

  if (linkRows.isEmpty) return result;

  final synergyIds = {...linkRows.map((r) => r.synergyId)}.toList();
  final rows = await (db.select(db.synergies)
        ..where((t) => t.userId.equals(userId) & t.id.isIn(synergyIds)))
      .get();
  final withLinks = await _rowsToSynergies(db, rows);
  final byId = {for (final s in withLinks) s.id: s};

  final seenPerHash = <int, Set<String>>{};
  for (final link in linkRows) {
    final hash = link.itemHash;
    if (hash == null) continue;
    final syn = byId[link.synergyId];
    if (syn == null) continue;
    final seen = seenPerHash.putIfAbsent(hash, () => <String>{});
    if (!seen.add(syn.id)) continue;
    result.putIfAbsent(hash, () => <SynergyWithLinks>[]).add(syn);
  }
  return result;
}

